File: app/src/main/java/mods/anti/AntiEdit.java
package mods.anti;

import android.content.Context;

import mods.constants.Constants;
import mods.preference.QuickAccessPrefs;
import mods.utils.LogUtils;

import com.discord.models.message.Message;
import com.discord.utilities.textprocessing.node.EditedMessageNode;
import com.discord.utilities.time.ClockFactory;
import com.discord.utilities.time.TimeUtils;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

import b.a.t.b.a.a;
import mods.constants.PreferenceKeys;
import mods.preference.Prefs;
import mods.utils.FileLogger;
import mods.utils.LRUCache;
import mods.utils.StoreUtils;
import mods.view.PrependEditNode;

public class AntiEdit {

    private static final LRUCache<Long, List<String>> editedMessages = new LRUCache<>(250);

    public static void appendEdits(Context context, Message message, List<Object> list) {
        if (StoreUtils.isMessageEdited(message)) {
            // faster to use get and check for null
            // compared to containsKey() then get()
            List<String> edits = editedMessages.get(message.getId());

            if (edits != null) {
                List<String> copy = new ArrayList<>(edits);
                Collections.reverse(copy);

                for (String edit : copy) {
                    list.add(0, new PrependEditNode(context, edit));
                    list.add(1, new EditedMessageNode(context));
                    list.add(2, new a("\n"));
                }
            }
        }
    }

    public static void parseEditedMessage(Map<Long, Message> map, com.discord.api.message.Message newMessage) {
        if (map == null || !StoreUtils.isMessageEdited(newMessage)) return;

        String mode = Prefs.getString(PreferenceKeys.ANTI_EDIT_MODE, "Off");

        if (!"Off".equalsIgnoreCase(mode)) {

            Message oldMessage = map.get(newMessage.o());

            if (oldMessage == null) {
                LogUtils.log("AntiEdit", "edited message not found in cache:\n" + newMessage.toString());
                return;
            }

            String previousContent = oldMessage.getContent();
            String newContent = newMessage.i();

            LogUtils.log("AntiEdit", "edited message found:\n" + previousContent);

            synchronized (AntiEdit.class) {
                // faster to use get and check for null
                // compared to containsKey() then get()
                List<String> edits = editedMessages.get(newMessage.o());

                if (edits == null) {
                    edits = new ArrayList<>(5);
                    edits.add(previousContent);
                    editedMessages.put(newMessage.o(), edits);
                } else {
                    if (edits.size() >= 5) {
                        edits.remove(0);
                    }
                    edits.add(previousContent);
                }
            }

            if ("Show Edit + Log".equalsIgnoreCase(mode)) {
                FileLogger.writeWithProfileInfo(
                        new Message(newMessage),
                        "messages",
                        "'" + previousContent + "' was changed to '" + newContent + "'",
                        "Edited Messages",
                        "edited"
                );
            }
        }
    }

    public static String getEditedStringWithTimestamp(Message message, Context context) {
        if (!QuickAccessPrefs.isEditTimestampEnabled() || message.getEditedTimestamp() == null) return EditedMessageNode.Companion.getEditedString(context);

        return " (" +
                context.getString(Constants.STRING_MESSAGE_EDITED) + ' ' +
                TimeUtils.toReadableTimeString(context, message.getEditedTimestamp().g(), ClockFactory.get()) +
                ')';
    }

    /*public static boolean parseEditedMessage(Map<Long, Message> map, Message newMessage, boolean editFound) {
        if (map == null || !StoreUtils.isMessageEdited(newMessage)) return editFound;

        String mode = Prefs.getString(PreferenceKeys.ANTI_EDIT_MODE, "Off");

        if (!"Off".equalsIgnoreCase(mode)) {

            Message oldMessage = map.get(newMessage.r());

            if (oldMessage == null) {
                LogUtils.log("AntiEdit", "edited message not found in cache:\n" + newMessage.toString());
            } else {
                editFound = true;

                String previousContent = oldMessage.k();
                String newContent = newMessage.k();

                LogUtils.log("AntiEdit", "edited message found:\n" + previousContent);

                synchronized (AntiEdit.class) {
                    // faster to use get and check for null
                    // compared to containsKey() then get()
                    List<String> edits = editedMessages.get(newMessage.r());

                    if (edits == null) {
                        edits = new ArrayList<>(5);
                        edits.add(previousContent);
                        editedMessages.put(newMessage.r(), edits);
                    } else {
                        if (edits.size() >= 5) {
                            edits.remove(0);
                        }
                        edits.add(previousContent);
                    }
                }

                if ("Show Edit + Log".equalsIgnoreCase(mode)) {
                    FileLogger.writeWithProfileInfo(newMessage,
                            "messages",
                            "'" + previousContent + "' was changed to '" + newContent + "'",
                            "Edited Messages",
                            "edited");
                }
            }
        }
        return editFound;
    }

    private static SpannableStringBuilder createFromEditedMessage(String content) {
        SpannableStringBuilder sb = new SpannableStringBuilder();
        sb.append(content);
        sb.append(" ");
        sb.append("(edited)");
        sb.append("\n\n");
        sb.setSpan(new ForegroundColorSpan(ColorCompat.getThemedColor(DiscordTools.getContext(), 0x7f040176)),
                0,
                content.length(),
                Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
        sb.setSpan(new RelativeSizeSpan(0.75f),
                content.length() - 10, // length of (edited) + 2 new lines
                content.length() - 2,  // do not span new lines
                Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
        return sb;
    }

    public static DraweeSpanStringBuilder appendEditedMessages(ModelMessage message, DraweeSpanStringBuilder sb) {
        if (message.messageEditHistory != null && !message.messageEditHistory.isEmpty()) {
            DraweeSpanStringBuilder sb2 = new DraweeSpanStringBuilder();
            for (String edit : message.messageEditHistory) {
                sb2.append(createFromEditedMessage(edit));
            }
            sb2.append(sb);
            return sb2;
        }
        return sb;
    }

    public static String appendEditedMessages(ModelMessage message) {
        if (message.messageEditHistory != null && !message.messageEditHistory.isEmpty()) {
            StringBuilder sb = new StringBuilder();
            for (String edit : message.messageEditHistory) {
                sb.append(edit);
                sb.append(" (edited)\n\n");
            }
            sb.append(message.getContent().trim());
            LogUtils.log("Bluecord", "after edit = " + sb.toString());
            return sb.toString();
        }
        return message.getContent();
    }

    */
}

File: app/src/main/java/mods/anti/AntiEdit.kt
package mods.anti

import android.content.Context
import b.a.t.b.a.a
import com.bluecord.R
import com.discord.models.message.Message
import com.discord.utilities.textprocessing.node.EditedMessageNode
import com.discord.utilities.time.ClockFactory
import com.discord.utilities.time.TimeUtils
import mods.db.Databases
import mods.db.EditedMessagesTable
import mods.preference.EditMode
import mods.preference.QuickAccessPrefs
import mods.preference.QuickAccessPrefs.isEditTimestampEnabled
import mods.promise.runCatchingOrLog
import mods.utils.FileLogger.writeWithProfileInfo
import mods.utils.LRUCache
import mods.utils.LogUtils
import mods.utils.StoreUtils
import mods.view.PrependEditNode
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

object AntiEdit {

    private val TAG = AntiEdit::class.java.simpleName
    private const val LIMIT_PER_MESSAGE = 5

    private val editedMessages = LRUCache<Long, ArrayDeque<EditedMessagesTable.EditedMessage>>(256)
    private val editAppendLock = ReentrantLock()

    @JvmStatic
    fun preloadEdits(messages: List<com.discord.api.message.Message>) {
        preloadEditsByIds(messages.map { it.id })
    }

    @JvmStatic
    fun preloadEditsFromStore(messages: List<Message>) {
        preloadEditsByIds(messages.map { it.id })
    }

    private fun preloadEditsByIds(messageIds: List<Long>) {
        if (messageIds.isEmpty()) {
            return
        }

        runCatchingOrLog {
            editAppendLock.withLock {
                val edits = Databases.get().edited.preloadEditedMessages(messageIds)
                val editsByMessageId = edits.groupBy { it.messageId }
                for ((id, messageEdits) in editsByMessageId) {
                    editedMessages.remove(id)
                    for (edit in messageEdits) {
                        putEditInCache(edit)
                    }
                }
            }
        }
    }

    private fun putEditInCache(edit: EditedMessagesTable.EditedMessage) {
        var edits = editedMessages.get(edit.messageId)
        if (edits == null) {
            edits = ArrayDeque(LIMIT_PER_MESSAGE)
            editedMessages.put(edit.messageId, edits)
        }
        while (edits.size >= LIMIT_PER_MESSAGE) {
            edits.removeFirst()
        }
        edits.addLast(edit)
    }

    @JvmStatic
    fun appendEdits(context: Context?, message: Message, nodes: List<Any>) {
        if (nodes !is ArrayList<Any>) return
        if (!StoreUtils.isMessageEdited(message)) return

        // faster to use get and check for null
        // compared to containsKey() then get()
        val edits = editedMessages.get(message.id)
        if (edits != null) {
            editAppendLock.withLock {
                for (edit in edits.asReversed()) {
                    nodes.add(0, PrependEditNode(context, edit.text))
                    nodes.add(1, EditedMessageNode(context))
                    nodes.add(2, a("\n"))
                }
            }
        }
    }

    @JvmStatic
    fun parseEditedMessage(
        map: MutableMap<Long, Message>?,
        newMessage: com.discord.api.message.Message?
    ) {
        if (map == null || newMessage == null) return
        if (!StoreUtils.isMessageEdited(newMessage)) return

        when (val mode = QuickAccessPrefs.editMode) {
            EditMode.OFF -> {}
            EditMode.ON, EditMode.ON_AND_LOG -> {
                val oldMessage = map[newMessage.id] ?: run {
                    LogUtils.log(TAG, "edited message not found in cache:\n$newMessage")
                    return@parseEditedMessage
                }
                LogUtils.log(TAG, "edited message found:\n${oldMessage.content}")

                val previousContent = oldMessage.content
                val newContent = newMessage.content
                val edit = EditedMessagesTable.EditedMessage(
                    channelId = oldMessage.channelId,
                    messageId = oldMessage.id,
                    text = oldMessage.content,
                    timestamp = StoreUtils.getServerSyncedTime(),
                )

                editAppendLock.withLock {
                    putEditInCache(edit)
                    Databases.get().edited.putEditedMessage(edit)
                }

                if (mode == EditMode.ON_AND_LOG) {
                    writeWithProfileInfo(
                        message = Message(newMessage),
                        type = "messages",
                        info = "'$previousContent' was changed to '$newContent'",
                        dir = "Edited Messages",
                        action = "edited"
                    )
                }
            }
        }
    }

    @JvmStatic
    fun getEditedStringWithTimestamp(message: Message, context: Context): String {
        if (!isEditTimestampEnabled || message.editedTimestamp == null) {
            return " (" + context.getString(R.string.message_edited) + ")"
        }
        return buildString {
            append(" (")
            append(context.getString(R.string.message_edited))
            append(' ')
            append(TimeUtils.toReadableTimeString(context, message.editedTimestamp.g(), ClockFactory.get()))
            append(')')
        }
    }
}
